src/crud_ops.rs

```diff
@@ -143,6 +225,49 @@ impl Client {
         counter!(metrics::total_delete_ops).increment(1);
         with_retries!(self, self.delete_impl(path).await)
     }
+
+    async fn bulk_delete_impl(&self, paths: &Vec<Path>) -> crate::Result<Vec<(Path, String)>> {
+        let stream = stream::iter(paths.iter().map(|path| Ok(path.clone()))).boxed();
+        // Counter to keep track of the index of the path that failed to delete
+        let counter = Arc::new(AtomicUsize::new(0));
+        let bulk_failed_entries = self.store.delete_stream(stream)
+            .filter_map(|result| async {
+                let counter_clone = Arc::clone(&counter);
+                let index = counter_clone.fetch_add(1, Ordering::SeqCst);
+                match result {
+                    Ok(path) => {
+                        None
+                    },
+                    Err(e) => match e {
+                        // We treat not found as success because AWS S3 does not return an error
+                        // if the object does not exist
+                        object_store::Error::NotFound { path: _, source } => {
+                            None
+                        },
+                        _ => {
+                            Some((paths[index].clone(), e.to_string()))
```

Comment:
From the above comment about the return type we can get a `crate::Error` instead of a `object_store::Error` by converting it with `.into()` here (and naming the correct type in collect):
```suggestion
                            Some((paths[index].clone(), e.into()))
```